tests/unit/lms/resources/_js_config_test.py

```diff
@@ -22,10 +22,89 @@ def test_it_is_mutable(self, config):
+    def test_it(self, context, js_config):
+        js_config.config["mode"] = "foo"
+
+        js_config.enable_content_item_selection_mode(
+            "test_form_action", "test_form_fields"
+        )
 
         assert js_config.config["mode"] == "content-item-selection"
+        assert js_config.config["formAction"] == "test_form_action"
+        assert js_config.config["formFields"] == "test_form_fields"
+        assert js_config.config["googleClientId"] == "fake_client_id"
+        assert js_config.config["googleDeveloperKey"] == "fake_developer_key"
+        assert (
+            js_config.config["customCanvasApiDomain"]
+            == context.custom_canvas_api_domain
+        )
+        assert js_config.config["lmsUrl"] == context.lms_url
+        assert js_config.config["ltiLaunchUrl"] == "http://example.com/lti_launches"
+
+    def test_it_enables_the_lms_file_picker(self, js_config):
+        js_config.enable_content_item_selection_mode(
+            "test_form_action", "test_form_fields"
+        )
+
+        assert js_config.config["enableLmsFilePicker"] is True
+        assert js_config.config["courseId"] == "test_course_id"
+
+    def test_it_doesnt_enable_the_lms_file_picker_if_the_lms_isnt_Canvas(
+        self, js_config, pyramid_request
+    ):
+        pyramid_request.params["tool_consumer_info_product_family_code"] = "whiteboard"
+        del pyramid_request.params["custom_canvas_course_id"]
+
+        js_config.enable_content_item_selection_mode(
+            "test_form_action", "test_form_fields"
+        )
+
+        assert not js_config.config["enableLmsFilePicker"]
+        assert "courseId" not in js_config.config
+
+    def test_it_doesnt_enable_the_lms_file_picker_if_the_consumer_key_isnt_found_in_the_db(
+        self, ai_getter, js_config
+    ):
+        ai_getter.developer_key.side_effect = ConsumerKeyError()
+
+        js_config.enable_content_item_selection_mode(
+            "test_form_action", "test_form_fields"
+        )
+
+        assert not js_config.config["enableLmsFilePicker"]
+        assert "courseId" not in js_config.config
+
+    def test_it_doesnt_enable_the_lms_file_picker_if_we_dont_have_a_developer_key(
+        self, ai_getter, js_config
+    ):
+        ai_getter.developer_key.return_value = None
+
+        js_config.enable_content_item_selection_mode(
+            "test_form_action", "test_form_fields"
+        )
+
+        assert not js_config.config["enableLmsFilePicker"]
+        assert "courseId" not in js_config.config
+
+    def test_it_doesnt_enable_the_lms_file_picker_if_theres_no_custom_canvas_course_id(
+        self, js_config, pyramid_request
+    ):
+        del pyramid_request.params["custom_canvas_course_id"]
+
+        js_config.enable_content_item_selection_mode(
+            "test_form_action", "test_form_fields"
+        )
+
+        assert not js_config.config["enableLmsFilePicker"]
+        assert "courseId" not in js_config.config
```

Comment:
Gah. I don't know how you'd do it exactly, but the duplication in the four above kind of kills me. I can't see how you'd parameterise it. Maybe an `_assert_lms_file_picker(js_config, enabled=False)` function instead? Then all of these tests would become some variation of:

```python
del pyramid_request.params["custom_canvas_course_id"]

self._assert_lms_file_picker(js_config, enabled=False)
```

Reply:
I think one of those `method_caller()` fixtures would do it, I'll add that.

Reply:
Oh, and for the two duplicated asserts an `asser_lms_file_picker_not_enabled()` fixture will work. If you make it a pytest fixture then you don't need to pass `js_config` to it 'cos it can itself depend on `js_config` as a fixture. It ends up meaning that the tests themselves don't need `js_config` as a parameter anymore

Reply:
Did https://github.com/hypothesis/lms/pull/1546/commits/e87eb93e2841b0bb54069afe7fa45b0a92fb4875 and https://github.com/hypothesis/lms/pull/1546/commits/294852dada5039c3946100bd071cdd8850698c3f. End result:

```python
class TestEnableContentItemSelectionMode:
    def test_it(self, context, js_config, enable_content_item_selection_mode):
        js_config.config["mode"] = "foo"

        enable_content_item_selection_mode()

        ... (lots of asserts)

    def test_it_doesnt_enable_the_lms_file_picker_if_the_lms_isnt_Canvas(
        self,
        context,
        enable_content_item_selection_mode,
        assert_lms_file_picker_not_enabled,
    ):
        context.is_canvas = False

        enable_content_item_selection_mode()

        assert_lms_file_picker_not_enabled()

    def test_it_doesnt_enable_the_lms_file_picker_if_the_consumer_key_isnt_found_in_the_db(
        self,
        ai_getter,
        enable_content_item_selection_mode,
        assert_lms_file_picker_not_enabled,
    ):
        ai_getter.developer_key.side_effect = ConsumerKeyError()

        enable_content_item_selection_mode()

        assert_lms_file_picker_not_enabled()

    def test_it_doesnt_enable_the_lms_file_picker_if_we_dont_have_a_developer_key(
        self,
        ai_getter,
        enable_content_item_selection_mode,
        assert_lms_file_picker_not_enabled,
    ):
        ai_getter.developer_key.return_value = None

        enable_content_item_selection_mode()

        assert_lms_file_picker_not_enabled()

    def test_it_doesnt_enable_the_lms_file_picker_if_theres_no_custom_canvas_course_id(
        self,
        pyramid_request,
        enable_content_item_selection_mode,
        assert_lms_file_picker_not_enabled,
    ):
        del pyramid_request.params["custom_canvas_course_id"]

        enable_content_item_selection_mode()

        assert_lms_file_picker_not_enabled()

    @pytest.fixture
    def assert_lms_file_picker_not_enabled(self, js_config):
        def assert_lms_file_picker_not_enabled():
            assert not js_config.config["enableLmsFilePicker"]
            assert "courseId" not in js_config.config

        return assert_lms_file_picker_not_enabled

    @pytest.fixture
    def enable_content_item_selection_mode(self, js_config):
        """
        Return a wrapper method for enable_content_item_selection_mode.

        Passing standard test arguments to the method.
        """

        def enable_content_item_selection_mode():
            js_config.enable_content_item_selection_mode(
                mock.sentinel.form_action, mock.sentinel.form_fields
            )

        return enable_content_item_selection_mode
```

Reply:
Hmm, I don't think this is better. We don't have to pass `js_config` into the tests, but seeing as these are tests about `js_config`, I'm not sure that's good. It's also longer than it was before.

Pretty much all of that additional complexity has been introduced to avoid passing a parameter. I think just because we can do this with `pytest` doesn't mean we should.

As an example here is the last test (with it's fixtures) in the current style:

```python
def test_it_doesnt_enable_the_lms_file_picker_if_theres_no_custom_canvas_course_id(
        self,
        pyramid_request,
        enable_content_item_selection_mode,
        assert_lms_file_picker_not_enabled,
    ):
        del pyramid_request.params["custom_canvas_course_id"]

        enable_content_item_selection_mode()

        assert_lms_file_picker_not_enabled()

    @pytest.fixture
    def assert_lms_file_picker_not_enabled(self, js_config):
        def assert_lms_file_picker_not_enabled():
            assert not js_config.config["enableLmsFilePicker"]
            assert "courseId" not in js_config.config

        return assert_lms_file_picker_not_enabled

    @pytest.fixture
    def enable_content_item_selection_mode(self, js_config):
        """
        Return a wrapper method for enable_content_item_selection_mode.

        Passing standard test arguments to the method.
        """

        def enable_content_item_selection_mode():
            js_config.enable_content_item_selection_mode(
                mock.sentinel.form_action, mock.sentinel.form_fields
            )

        return enable_content_item_selection_mode
```

Here is the same thing with the condition inline and with a plain function:

```python
    def test_it_doesnt_enable_the_lms_file_picker_if_theres_no_custom_canvas_course_id(
        self, pyramid_request, js_config,
    ):
        del pyramid_request.params["custom_canvas_course_id"]

        js_config.enable_content_item_selection_mode(
            mock.sentinel.form_action, mock.sentinel.form_fields
        )

        self.assert_lms_file_picker_not_enabled(js_config)

    def assert_lms_file_picker_not_enabled(self, js_config):
        assert not js_config.config["enableLmsFilePicker"]
        assert "courseId" not in js_config.config
```

There's just a lot less to read and understand there.

I was originally thinking to move the setup into the function too:

```python
    def test_it_doesnt_enable_the_lms_file_picker_if_theres_no_custom_canvas_course_id(
        self, pyramid_request, js_config,
    ):
        del pyramid_request.params["custom_canvas_course_id"]

        self.assert_lms_file_picker_not_enabled(js_config)

    def assert_lms_file_picker_not_enabled(self, js_config):
        js_config.enable_content_item_selection_mode(
            mock.sentinel.form_action, mock.sentinel.form_fields
        )

        assert not js_config.config["enableLmsFilePicker"]
        assert "courseId" not in js_config.config
```

This would lead to very short tests over all, but it looses the setup / do something / assert split, which is quite nice.

Reply:
Ah ok, I've gone ahead and done this: https://github.com/hypothesis/lms/pull/1546/commits/89b7fe47fef64db96a4829a05e9e1c7355df7fca